File: tests/test_ch157_regression_metrics.py
"""Tests for aiinaction.ch157_regression_metrics, including shared fixtures.

The fixtures in EXPECTED are the single source of truth: the Julia and Rust test
suites assert against the same numbers, which keeps the three implementations at
parity. The residual fixture is the classic four-point set also used by the core
``metrics`` parity tests.
"""
from __future__ import annotations

import pytest

from aiinaction import ch157_regression_metrics as rm

# Shared fixtures mirrored in julia/AIInAction/test and rust/aiinaction/src.
Y_TRUE = [3.0, -0.5, 2.0, 7.0]
Y_PRED = [2.5, 0.0, 2.0, 8.0]
DELTA = 0.75
EXPECTED = {
    "mse": 0.375,
    "rmse": 0.6123724356957945,
    "mae": 0.5,
    # Per-observation Huber loss with delta=0.75 on the residuals [0.5, -0.5, 0, -1].
    "huber_each": [0.125, 0.125, 0.0, 0.46875],
    "huber_mean": 0.1796875,
}


def test_mse_matches_fixture():
    assert rm.mse(Y_TRUE, Y_PRED) == pytest.approx(EXPECTED["mse"])


def test_rmse_matches_fixture():
    assert rm.rmse(Y_TRUE, Y_PRED) == pytest.approx(EXPECTED["rmse"])


def test_mae_matches_fixture():
    assert rm.mae(Y_TRUE, Y_PRED) == pytest.approx(EXPECTED["mae"])


def test_rmse_is_sqrt_mse():
    assert rm.rmse(Y_TRUE, Y_PRED) == pytest.approx(rm.mse(Y_TRUE, Y_PRED) ** 0.5)


def test_huber_each_matches_fixture():
    assert rm.huber_loss(Y_TRUE, Y_PRED, delta=DELTA) == pytest.approx(
        EXPECTED["huber_each"]
    )


def test_huber_mean_matches_fixture():
    assert rm.huber_loss_mean(Y_TRUE, Y_PRED, delta=DELTA) == pytest.approx(
        EXPECTED["huber_mean"]
    )


def test_huber_continuity_at_threshold():
    # At |r| == delta the quadratic and linear branches agree: both give 0.5*delta^2.
    d = 1.5
    quad = rm.huber_loss([0.0], [d], delta=d)[0]
    lin = rm.huber_loss([0.0], [d + 1e-9], delta=d)[0]
    assert quad == pytest.approx(0.5 * d * d)
    assert lin == pytest.approx(0.5 * d * d, abs=1e-6)


def test_huber_large_delta_approaches_half_mse():
    # As delta -> inf every residual is an inlier, so mean Huber -> 0.5 * MSE.
    big = rm.huber_loss_mean(Y_TRUE, Y_PRED, delta=1000.0)
    assert big == pytest.approx(0.5 * rm.mse(Y_TRUE, Y_PRED))


def test_perfect_prediction_is_zero():
    assert rm.mse(Y_TRUE, Y_TRUE) == 0.0
    assert rm.rmse(Y_TRUE, Y_TRUE) == 0.0
    assert rm.mae(Y_TRUE, Y_TRUE) == 0.0
    assert rm.huber_loss_mean(Y_TRUE, Y_TRUE, delta=1.0) == 0.0


@pytest.mark.parametrize("fn", [rm.mse, rm.huber_loss, rm.huber_loss_mean])
def test_length_mismatch_raises(fn):
    with pytest.raises(ValueError, match="length mismatch"):
        fn([1.0, 2.0], [1.0])


@pytest.mark.parametrize("fn", [rm.mse, rm.huber_loss, rm.huber_loss_mean])
def test_empty_raises(fn):
    with pytest.raises(ValueError, match="non-empty"):
        fn([], [])


@pytest.mark.parametrize("bad_delta", [0.0, -1.0, float("inf"), float("nan")])
def test_huber_bad_delta_raises(bad_delta):
    with pytest.raises(ValueError, match="delta must be a positive"):
        rm.huber_loss(Y_TRUE, Y_PRED, delta=bad_delta)

File: julia/AIInAction/test/test_ch157_regression_metrics.jl
using Test
using AIInAction.Ch157RegressionMetrics

# Shared fixtures: identical to the Python and Rust test suites. Every top-level
# const is prefixed with RM157 so it cannot collide with other test files.
const RM157_Y_TRUE = [3.0, -0.5, 2.0, 7.0]
const RM157_Y_PRED = [2.5, 0.0, 2.0, 8.0]
const RM157_DELTA = 0.75

@testset "Ch157 regression-metrics parity fixtures" begin
    @test mse(RM157_Y_TRUE, RM157_Y_PRED) ≈ 0.375
    @test rmse(RM157_Y_TRUE, RM157_Y_PRED) ≈ 0.6123724356957945
    @test mae(RM157_Y_TRUE, RM157_Y_PRED) ≈ 0.5
    @test huber_loss(RM157_Y_TRUE, RM157_Y_PRED; delta=RM157_DELTA) ≈
          [0.125, 0.125, 0.0, 0.46875]
    @test huber_loss_mean(RM157_Y_TRUE, RM157_Y_PRED; delta=RM157_DELTA) ≈ 0.1796875
end

@testset "Ch157 regression-metrics properties and edge cases" begin
    @test rmse(RM157_Y_TRUE, RM157_Y_PRED) ≈ sqrt(mse(RM157_Y_TRUE, RM157_Y_PRED))

    # Continuity at the threshold: both branches give 0.5*delta^2.
    d = 1.5
    @test huber_loss([0.0], [d]; delta=d)[1] ≈ 0.5 * d * d
    @test huber_loss([0.0], [d + 1e-9]; delta=d)[1] ≈ 0.5 * d * d atol = 1e-6

    # Large delta: mean Huber -> 0.5 * MSE.
    @test huber_loss_mean(RM157_Y_TRUE, RM157_Y_PRED; delta=1000.0) ≈
          0.5 * mse(RM157_Y_TRUE, RM157_Y_PRED)

    # Perfect prediction is zero everywhere.
    @test mse(RM157_Y_TRUE, RM157_Y_TRUE) == 0.0
    @test rmse(RM157_Y_TRUE, RM157_Y_TRUE) == 0.0
    @test huber_loss_mean(RM157_Y_TRUE, RM157_Y_TRUE; delta=1.0) == 0.0

    @test_throws ArgumentError mse([1.0, 2.0], [1.0])
    @test_throws ArgumentError huber_loss([1.0, 2.0], [1.0]; delta=1.0)
    @test_throws ArgumentError mse(Float64[], Float64[])
    @test_throws ArgumentError huber_loss(RM157_Y_TRUE, RM157_Y_PRED; delta=0.0)
    @test_throws ArgumentError huber_loss(RM157_Y_TRUE, RM157_Y_PRED; delta=-1.0)
    @test_throws ArgumentError huber_loss(RM157_Y_TRUE, RM157_Y_PRED; delta=Inf)
end
